flow/controllers/base_controller.py

```diff
@@ -107,16 +121,46 @@ def get_action(self, env):
         if accel is None:
             return None
 
+        # store the acceleration without noise to each vehicle
+        # run fail safe if requested
+        env.k.vehicle.update_accel(self.veh_id, accel, noise=False, failsafe=False)
+        accel_no_noise_with_failsafe = accel
+
+        if self.fail_safe is not None:
+            for check in self.fail_safe:
+                if check == 'instantaneous':
+                    accel_no_noise_with_failsafe = self.get_safe_action_instantaneous(
+                        env, accel_no_noise_with_failsafe)
+                elif check == 'safe_velocity':
+                    accel_no_noise_with_failsafe = self.get_safe_velocity_action(
+                        env, accel_no_noise_with_failsafe)
+                elif check == 'feasible_accel':
+                    accel_no_noise_with_failsafe = self.get_feasible_action(
+                        accel_no_noise_with_failsafe)
+                elif check == 'obey_speed_limit':
+                    accel_no_noise_with_failsafe = self.get_obey_speed_limit_action(
+                        env, accel_no_noise_with_failsafe)
```

Comment:
Maybe do a mapping in `__init__` like

```python
failsafes = { 'instantaneous': self.get_safe_action_instantaneous, 
              'safe_velocity': ... }
```

and then

```python
if self.fail_safe is not None:
    for check in self.fail_safe:
        if check in self.failsafes:
            accel_no_noise_with_failsafe = self.failsafes[check](env, accel_no_noise_with_failsafe)
```

also the `in` could be checked in the constructor so that it's not checked at every iteration:

```python
if isinstance(fail_safe, str):
    self.fail_safe = [fail_safe]
elif isinstance(fail_safe, list) or fail_safe is None:
    self.fail_safe = fail_safe
else:
    raise ValueError("fail_safe should be string or list of strings.\nSet fal_safe to None\n")
    self.fail_safe = None

for check in self.fail_safe:
    if check not in self.failsafes:
        raise ValueError(f'{check} is not a valid failsafe')
```